### PR TITLE
Dockerized email inbound parse webhook

### LINKED ISSUE #650: Create a Dockerized email inbound parse webhook consumer
#### Issue Summary

SendGrid has an [Email Inbound Parse Webhook](https://sendgrid.com/docs/API_Reference/Parse_Webhook/inbound_email.html) which posts data from a specified incoming email address to a URL of your choice. We want to have an easily deployable solution that allows for customers to easiy get up and running processing (parse and complete some action) their inbound parse webhooks.

#### Acceptance Criteria

* A folder which contains a Dockerized solution
* Example documentation on how to deploy locally and start processing incoming emails
* Example documentation on how to deploy on at least one cloud service and start processing incoming emails
* An example complete payload with a test to verify the data is parsed properly

### PR DESCRIPTION
**Pull request by ItsMeWithTheFace**
# Fixes #650 # 

### Checklist
- [x] I have made a material change to the repo (functionality, testing, spelling, grammar)
- [x] I have read the [Contribution Guide] and my PR follows them.
- [x] I updated my branch with the master branch.
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] I have added necessary documentation about the functionality in the appropriate .md file
- [x] I have added inline documentation to the code I modified

### Short description of what this PR does:
- A small python 3 Flask application that parses inbound email data and prints it out
- README guides for local deployment and deployment on to GCP

If you have questions, please send an email to [SendGrid](mailto:dx@sendgrid.com), or file a GitHub Issue in this repository.

### CODE REVIEW

**Comment by SendGridDX** [comment]
[CLA assistant check](https://cla.sendgrid.com/sendgrid/sendgrid-python?pullRequest=716) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.sendgrid.com/sendgrid/sendgrid-python?pullRequest=716) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.sendgrid.com/check/sendgrid/sendgrid-python?pullRequest=716) it.</sub>

**Comment by ItsMeWithTheFace** [comment]
@thinkingserious I added steps to deploy to Google Kubernetes Engine if that's alright. Let me know if there are any issues.

File: examples/inbound-webhook-parsing/example-webhook-payload.txt [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+
+----
+Content For: sender_ip
+
+209.85.210.179
+
+----
+Content For: envelope
+
+{"to":["test@rakinu.com"],"from":"rakinuddin97@gmail.com"}
```
Comment (misterdorm) [inline examples/inbound-webhook-parsing/example-webhook-payload.txt]: These addresses appear throughout the example data.  Are you comfortable with these being here in this public repo?

Comment (misterdorm) [reply examples/inbound-webhook-parsing/example-webhook-payload.txt]: Yes, that's totally fine.

**Comment by misterdorm** [comment]
Would you please also rebase this against the v4 branch?  Thanks!

**Comment by ItsMeWithTheFace** [comment]
@misterdorm Changes have been made. Let me know if there's anything else :)

File: examples/inbound-webhook-parsing/example-webhook-payload.txt [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+X-Google-DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed; d=1e100.net; s=20161025; h=x-gm-message-state:mime-version:references:in-reply-to:from:date :message-id:subject:to; bh=51n5icMQdyoO44ELkOW/wBx5yqtLMQENLcy1hVc+id8=; b=OHdk1lTtqDWm0bCYZ+BubTUSByOjiJ1j7DxQ/bFP0G/ig7yfsHvuCc2kjwRYDzpG62 OWvDvwh1qv5mlfmSHvSVcSSg1vQCuH/tRtDoUn/x2lWwdABBQjN8IwJs2UCNXMIRrmfF e6aZ4QZs3PeXst4sdyfifXvNa3Tkp1je/aGLU+3oO96ETlbbgL+f1UHK/BYgWkUwOLv9 TwMZBJ3k9lfvwrX20XwLQHMO/swR4xySx+NaQtI5Vb/P6ROr/QJg3HmPnVvyvW6aMeUC Mg4Jcae6hT176xbqoA8ED9sdVCnfXxdiPPoHwSAIuFPBNTWxJsyas6tJcFCfWJbdTbCP Fyfg==
+X-Gm-Message-State: AGRZ1gKm7xaY2ME2Ki6DyLeVVf77KXXecEtBOJuR1YoHCG9l+Qlcl8Rt jqDnY7stpDWsfs8NWq9+sMvmcLH6mnuOjJo7EhFETw9o
+X-Google-Smtp-Source: AJdET5eKfRS5sqL6309QMKzt4Ro8Rw5fhH2kS3sCk5vBsKSvxPqr8g/XSQe0Ak3aFQbtZ4LyBx7B6cwHv7kzLVOqoYw=
+X-Received: by 2002:a63:ce56:: with SMTP id r22-v6mr11536811pgi.217.1540771074645; Sun, 28 Oct 2018 16:57:54 -0700 (PDT)
+MIME-Version: 1.0
+References: <CAAzJDmJhGrJS9zc00U0EP6-mw+rc2MGX51ZSSUMEoYBiM=et4g@mail.gmail.com>
+In-Reply-To: <CAAzJDmJhGrJS9zc00U0EP6-mw+rc2MGX51ZSSUMEoYBiM=et4g@mail.gmail.com>
+From: Rakin Uddin <rakinuddin97@gmail.com>
```
Comment (misterdorm) [inline examples/inbound-webhook-parsing/example-webhook-payload.txt]: One last place where real email address is.

Comment (ItsMeWithTheFace) [reply examples/inbound-webhook-parsing/example-webhook-payload.txt]: Haha I got it now

**Comment by ItsMeWithTheFace** [comment]
@thinkingserious @misterdorm  I think this is all good to go now. Is it alright to merge?

**Comment by thinkingserious** [comment]
Hello @ItsMeWithTheFace,<br /><br/>Thanks again for the PR!<br />                                              <br />It's <a href="https://sendgrid.com/blog/hacktoberfest-2018-has-arrived/">HACKTOBERFEST</a>! We want to show our appreciation by sending                                              you some special Hacktoberfest swag. If you have not already, could you please fill out                                              <a href="https://dx.sendgrid.com/hacktoberfest/join">this form</a> so we can send it to you? Thanks!                                             <br /><br />Team SendGrid DX

File: examples/inbound-webhook-parsing/app.py [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+    output += "[BODY]\n"
+    output += data['text']
+    output += "\n\n"
+    output += "[ATTACHMENTS]\n"
+    output += data['attachments']
+    output += "\n\n"
+    output += "[CHARSETS]\n"
+    output += data['charsets']
+    output += "\n\n"
+    output += "------------------ END WEBHOOK DATA ------------------"
```
Comment (42B) [inline examples/inbound-webhook-parsing/app.py]: It seems slightly verbose to create a multi-line string this way? I know the README file above mentions Python 3.6+ which could give the option of using f-strings but even just `str.format` might be able to clean it up a bit?

```python
In[2]: import textwrap
  ...: 
  ...: data = {
  ...:     'headers': 'headers',
  ...:     'dkim': 'dkim',
  ...:     'from': 'from',
  ...:     'to': 'to',
  ...:     'subject': 'subject',
  ...:     'text': 'text',
  ...:     'attachments': 'attachments',
  ...:     'charsets': 'charsets'
  ...: }
  ...: 
  ...: 
  ...: def inbound_parser():
  ...:     output = ""
  ...:     output += "------------------ INCOMING WEBHOOK DATA ------------------"
  ...:     output += "\n\n"
  ...:     output += "[HEADERS]\n"
  ...:     output += data['headers']
  ...:     output += "\n\n"
  ...:     output += "[DKIM]\n"
  ...:     output += data['dkim']
  ...:     output += "\n\n"
  ...:     output += "[FROM]\n"
  ...:     output += data['from']
  ...:     output += "\n\n"
  ...:     output += "[TO]\n"
  ...:     output += data['to']
  ...:     output += "\n\n"
  ...:     output += "[SUBJECT]\n"
  ...:     output += data['subject']
  ...:     output += "\n\n"
  ...:     output += "[BODY]\n"
  ...:     output += data['text']
  ...:     output += "\n\n"
  ...:     output += "[ATTACHMENTS]\n"
  ...:     output += data['attachments']
  ...:     output += "\n\n"
  ...:     output += "[CHARSETS]\n"
  ...:     output += data['charsets']
  ...:     output += "\n\n"
  ...:     output += "------------------ END WEBHOOK DATA ------------------"
  ...:     return output
  ...: 
  ...: 
  ...: def inbound_parser_fstring():
  ...:     output = textwrap.dedent(f"""\
  ...:         ------------------ INCOMING WEBHOOK DATA ------------------
  ...: 
  ...:         [HEADERS]
  ...:         {data['headers']}
  ...: 
  ...:         [DKIM]
  ...:         {data['dkim']}
  ...: 
  ...:         [FROM]
  ...:         {data['from']}
  ...: 
  ...:         [TO]
  ...:         {data['to']}
  ...: 
  ...:         [SUBJECT]
  ...:         {data['subject']}
  ...: 
  ...:         [BODY]
  ...:         {data['text']}
  ...: 
  ...:         [ATTACHMENTS]
  ...:         {data['attachments']}
  ...: 
  ...:         [CHARSETS]
  ...:         {data['charsets']}
  ...: 
  ...:         ------------------ END WEBHOOK DATA ------------------""")
  ...:     return output
  ...: 
  ...: 
  ...: def inbound_parser_format():
  ...:     output = textwrap.dedent("""\
  ...:         ------------------ INCOMING WEBHOOK DATA ------------------
  ...: 
  ...:         [HEADERS]
  ...:         {headers}
  ...: 
  ...:         [DKIM]
  ...:         {dkim}
  ...: 
  ...:         [FROM]
  ...:         {from}
  ...: 
  ...:         [TO]
  ...:         {to}
  ...: 
  ...:         [SUBJECT]
  ...:         {subject}
  ...: 
  ...:         [BODY]
  ...:         {text}
  ...: 
  ...:         [ATTACHMENTS]
  ...:         {attachments}
  ...: 
  ...:         [CHARSETS]
  ...:         {charsets}
  ...: 
  ...:         ------------------ END WEBHOOK DATA ------------------""".format(**data))
  ...:     return output
  ...: 
In[3]: inbound_parser() == inbound_parser_fstring() == inbound_parser_format()
Out[3]: True
```

Comment (ItsMeWithTheFace) [reply examples/inbound-webhook-parsing/app.py]: Yeah sure I can check it out

Comment (ItsMeWithTheFace) [reply examples/inbound-webhook-parsing/app.py]: Decided to go the str.format route for better compatibility. Let me know if there are any other issues.

**Pull request closed without merging**